### PR TITLE
SBOM generation into configmap on build

### PR DESCRIPTION
**Pull request by timclifford**
This PR adds a routine to the build-deply script that creates a ConfigMap containing a SBOM in the environment namespace for each service.

This SBOM is initially in the CycloneDX format to consume less storage, but we are working on alternatives.

This SBOM is generated by Syft directly in the docker-host, and triggered on any deployment, and should not need to be updated outside of a deployment, given the limited opportunities to update modules outside of a build.

The exact composition of the Kubernetes labels is still to be determines, as is whether the ConfigMap should contain any additional fields.

Screenshot-20220202151947-2155x826

### CODE REVIEW

File: images/kubectl-build-deploy-dind/build-deploy-docker-compose.sh [hunk 1]
```diff
@@ -1669,3 +1669,22 @@ currentStepEnd="$(date +"%Y-%m-%d %H:%M:%S")"
...
+set -x
+
+for IMAGE_NAME in "${!IMAGES_BUILD[@]}"
+do
+
+  IMAGE_TAG="${IMAGE_TAG:-latest}"
+  IMAGE_FULL="docker:${REGISTRY}/${PROJECT}/${ENVIRONMENT}/${IMAGE_NAME}:${IMAGE_TAG}"
+
+  . /kubectl-build-deploy/scripts/exec-generate-sbom-configmap.sh
+done
```
Comment (tobybellwood) [inline images/kubectl-build-deploy-dind/build-deploy-docker-compose.sh]:
```suggestion
##############################################
### RUN sbom generation and store in configmap
##############################################

for IMAGE_NAME in "${!IMAGES_BUILD[@]}"
do

  IMAGE_TAG="${IMAGE_TAG:-latest}"
  IMAGE_FULL="${REGISTRY}/${PROJECT}/${ENVIRONMENT}/${IMAGE_NAME}:${IMAGE_TAG}"
  . /kubectl-build-deploy/scripts/exec-generate-sbom-configmap.sh
done

set +x
currentStepEnd="$(date +"%Y-%m-%d %H:%M:%S")"
patchBuildStep "${buildStartTime}" "${previousStepEnd}" "${currentStepEnd}" "${NAMESPACE}" "sbomCompleted" "SBOM Gathering"
previousStepEnd=${currentStepEnd}
set -x
```

File: images/kubectl-build-deploy-dind/scripts/exec-generate-sbom-configmap.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            --from-file=${TMP_DIR}/${REPO}-${IMAGE_NAME}.cyclonedx.json
+        kubectl --insecure-skip-tls-verify \
+            -n ${NAMESPACE} \
+            label configmap ${REPO}-${IMAGE_NAME}-sbom.config \
+            lagoon.sh=insights-sbom
+    fi
+    set +x
+else
+    echo "Failed to generate SBOM for ${IMAGE_FULL}"
+fi
```
Comment (tobybellwood) [inline images/kubectl-build-deploy-dind/scripts/exec-generate-sbom-configmap.sh]:
```suggestion
#!/bin/bash

# SBOM config
TMP_DIR="${TMP_DIR:-/tmp}"
SBOM_OUTPUT="cyclonedx-json"
SBOM_OUTPUT_FILE="${TMP_DIR}/${IMAGE_NAME}.cyclonedx.json"

set -x
# Run sbom and dump to file
echo "Running sbom scan using syft"
echo "Image being scanned: ${IMAGE_FULL}"
set +x

DOCKER_HOST=docker-host.lagoon.svc docker run --rm -v /var/run/docker.sock:/var/run/docker.sock imagecache.amazeeio.cloud/anchore/syft packages ${IMAGE_FULL} -o ${SBOM_OUTPUT} > ${SBOM_OUTPUT_FILE}

if [ -f "${SBOM_OUTPUT_FILE}" ]; then
    echo "Successfully generated SBOM for ${IMAGE_FULL}"

    SBOM_CONFIGMAP=lagoon-sbom-${IMAGE_NAME}

    set -x
    # If sbom configmap already exists then we need to update, else create new
    if kubectl --insecure-skip-tls-verify -n ${NAMESPACE} get configmap $SBOM_CONFIGMAP &> /dev/null; then
        kubectl --insecure-skip-tls-verify \
            -n ${NAMESPACE} \
            create configmap $SBOM_CONFIGMAP \
            --from-file=${SBOM_OUTPUT_FILE} \
            -o json \
            --dry-run=client | kubectl replace -f -
        kubectl --insecure-skip-tls-verify \
            -n ${NAMESPACE} \
            label configmap ${SBOM_CONFIGMAP} \
            lagoon.sh/insights=sbom \
            lagoon.sh/project=${PROJECT} \
            lagoon.sh/environment=${ENVIRONMENT} \
            lagoon.sh/service=${IMAGE_NAME}

    else
        # Create configmap and add label (#have to add label separately: https://github.com/kubernetes/kubernetes/issues/60295)
        kubectl --insecure-skip-tls-verify \
            -n ${NAMESPACE} \
            create configmap ${SBOM_CONFIGMAP} \
            --from-file=${SBOM_OUTPUT_FILE}
        kubectl --insecure-skip-tls-verify \
            -n ${NAMESPACE} \
            label configmap ${SBOM_CONFIGMAP} \
            lagoon.sh/insights=sbom \
            lagoon.sh/project=${PROJECT} \
            lagoon.sh/environment=${ENVIRONMENT} \
            lagoon.sh/service=${IMAGE_NAME}
    fi
    set +x
else
    echo "Failed to generate SBOM for ${IMAGE_FULL}"
fi
```

**Comment by tobybellwood** [comment]
ok worked this out locally:

Added a coupla chunky comments - but basically:
Using the docker-host to run the anchor/syft container allows it to also inspect other containers on docker-host
```
DOCKER_HOST=docker-host.lagoon.svc docker run --rm -v /var/run/docker.sock:/var/run/docker.sock imagecache.amazeeio.cloud/anchore/syft packages ${IMAGE_FULL} -o ${SBOM_OUTPUT} > ${SBOM_OUTPUT_FILE}
```
I've made a few naming changes too, just to make it a bit easier to relate to files/configmaps

I've also added some more standard labels onto the configMap to pick up with the controller

I've tested with an elasticsearch image (traditionally our largest SBOM) and it took ~20 seconds to generate and store the pair of configmaps
image

File: images/kubectl-build-deploy-dind/scripts/exec-generate-sbom-configmap.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            -n ${NAMESPACE} \
+            create configmap $SBOM_CONFIGMAP \
+            --from-file=${SBOM_OUTPUT_FILE} \
+            -o json \
+            --dry-run=client | kubectl replace -f -
+        kubectl --insecure-skip-tls-verify \
+            -n ${NAMESPACE} \
+            label configmap ${SBOM_CONFIGMAP} \
+            lagoon.sh/insights=sbom \
+            lagoon.sh/project=${PROJECT} \
```
Comment (timclifford) [inline images/kubectl-build-deploy-dind/scripts/exec-generate-sbom-configmap.sh:34]: lagoon.sh/insights-project

Comment (timclifford) [reply images/kubectl-build-deploy-dind/scripts/exec-generate-sbom-configmap.sh:34]: all labels to be  prefixed with 'insights'

Comment (tobybellwood) [reply images/kubectl-build-deploy-dind/scripts/exec-generate-sbom-configmap.sh:34]: Will keep project/service/environment as usual lagoon labels, but have made insightsType to denote sbom/drutiny etc

**Pull request merged**